core/src/main/java/feign/RequestTemplate.java

```diff
@@ -215,7 +215,8 @@ public RequestTemplate resolve(Map<String, ?> unencoded) {
     for (String field : headers.keySet()) {
       Collection<String> resolvedValues = new ArrayList<String>();
       for (String value : valuesOrEmpty(headers, field)) {
-        String resolved = urlDecode(expand(value, encoded));
+        String expandedValueWithEncodedPlus = expand(value, encoded).replaceAll("\\+", "__PLUS__");
```

Comment:
Actually it doesn't. decode just changes `+` into ` ` (space). I thought that it will be exactly what you wrote.